### ISSUE TITLE
ORA-01013 in v2.8.13

### ISSUE THREAD

**Issue by Tommi2Day** [sijms/go-ora#548]
My application states ORA-01013 errors when accessing  a prepared query. This error  doesn't exist with v2.8.11

**Comment by rpuchadm** [comment]
me too

**Comment by sijms** [comment]
fixed in next release

**Comment by sijms** [comment]
no same issue mentioned [here](https://github.com/sijms/go-ora/issues/545#issuecomment-2071874779)

**Comment by sijms** [comment]
fixed 2.8.14

**Comment by freeNestor** [comment]
> Just for understanding: You marked it as duplicate, but ORA-1013 "user requested cancel of current operation" is different error than ORA-1009 "missing mandantory parameter," isnt? Or is the same reason behind?

@sijms I think it's different. I have this issue too now after upgrading from 2.8.11.
I have a simple sql for watching db health every 15s, and context timeout is set to 5s. The new version 2.8.14 or 2.8.13 is unstable and there will be errors ORA-1013, but the old version has never been.
```
...
                                 selecttest:="SELECT systimestamp FROM dual"
                                 pingtmout, pingcle := context.WithTimeout(context.Background(), time.Second*time.Duration(5))
				getconn, err = srcdbtest.Connx(pingtmout)
				if err != nil {
					Zaplogger.Error("watchdog get conn failed", zap.String("driver", dstdriver),
						zap.String("dst-db-host", dsthost), zap.Int("dst-db-port", dstport),
						zap.String("dst-dbname", dstdbname), zap.Int("max-connection", maxConn), zap.Error(err))
				} else {
					_, err = getconn.ExecContext(pingtmout, selecttest)
					getconn.Close()
					if err != nil {
						Zaplogger.Error("watchdog exec failed", zap.String("driver", dstdriver),
							zap.String("dst-db-host", dsthost), zap.Int("dst-db-port", dstport),
							zap.String("dst-dbname", dstdbname), zap.Int("max-connection", maxConn), zap.Error(err))
					}
				}
...
```

**Comment by sijms** [comment]
I make a testing code
```golang
package main

import (
	"context"
	"database/sql"
	"fmt"
	go_ora "github.com/sijms/go-ora/v2"
	"os"
	"time"
)

func main() {
	config, err := go_ora.ParseConfig(os.Getenv("DSN"))
	if err != nil {
		fmt.Println("can't parse DSN:", err)
		return
	}
	go_ora.RegisterConnConfig(config)
	db, err := sql.Open("oracle", "")
	if err != nil {
		fmt.Println("can't open DB:", err)
		return
	}
	defer func() {
		err := db.Close()
		if err != nil {
			fmt.Println("can't close DB:", err)
		}
	}()
	pingtmout, pingcle := context.WithTimeout(context.Background(), time.Second*time.Duration(5))
	defer pingcle()
	_, err = db.ExecContext(pingtmout, "SELECT systimestamp from dual")
	if err != nil {
		fmt.Println("can't ping DB:", err)
		return
	}

	pingtimout, pingcancel := context.WithTimeout(context.Background(), time.Second*time.Duration(5))
	defer pingcancel()
	_, err = db.ExecContext(pingtimout, "SELECT systimestamp from dual")
	if err != nil {
		fmt.Println("can't ping 2 DB:", err)
	}
}
```
and result is ok

**Comment by freeNestor** [comment]
It's weird in my environment. Can you use below code to test? I also paste my environment info below.
go-ora: 2.8.14
go version to build: 1.22.2
build arch: linux-amd64
os: centos 7.8

```
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/jmoiron/sqlx"
	go_ora "github.com/sijms/go-ora/v2"
)

func main() {
	options := map[string]string{
		"TIMEOUT":            "60",
		"CONNECTION TIMEOUT": "5",
	}
	dsn := go_ora.BuildUrl(host, 1521, "orcldb", "test", "pass", options)

	conn, err := sqlx.Open("oracle", dsn)
	if err != nil {
		fmt.Printf("failed: %v\n", err)
		return
	}
	conn.SetMaxIdleConns(2)
	conn.SetMaxOpenConns(2)
	defer func() {
		err = conn.Close()
		if err != nil {
			fmt.Println("can't close db:", err)
		}
	}()

	selecttest := "SELECT systimestamp FROM dual"
	for {
		pingtmout, pingcle := context.WithTimeout(context.Background(), time.Second*time.Duration(5))
		getdur := time.Now().Local()
		getconn, err := conn.Connx(pingtmout)
		if err != nil {
			fmt.Println("can't get conn:", err)
		} else {
			execst := time.Now().Local()
			_, err = getconn.ExecContext(pingtmout, selecttest)
			getconn.Close()
			if err != nil {
				fmt.Println(time.Now().Local(), " ping db failed: ", err, " getdura: ", execst.Sub(getdur), " execdur: ", time.Since(execst))
			} else {
				fmt.Println(time.Now().Local(), " ping db succeeded, getdura: ", execst.Sub(getdur), " execdur: ", time.Since(execst))
			}
		}
		pingcle()
		time.Sleep(time.Second * 5)
	}
}
```

my test result is not ok:

<img width="1103" alt="image" src="https://github.com/sijms/go-ora/assets/19606302/00ca782f-0178-4150-ad10-7e52e247157e">

**Comment by sijms** [comment]
you use `Connection timeout = 60` and `Timeout = 5` they are same parameter

**Comment by sijms** [comment]
output
```golang
2024-04-28 15:34:19.424186 +0300 +03  ping db succeeded, getdura:  831.739ms  execdur:  98.159ms
2024-04-28 15:34:24.530289 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  104.891ms
2024-04-28 15:34:29.630483 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  98.743ms
2024-04-28 15:34:34.722144 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  90.692ms
2024-04-28 15:34:39.825108 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  102.512ms
2024-04-28 15:34:44.918666 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  93.249ms
2024-04-28 15:34:50.015942 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  96.829ms
2024-04-28 15:34:55.112127 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  95.9ms
2024-04-28 15:35:00.235455 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  122.728ms
2024-04-28 15:35:05.329036 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  93.175ms
2024-04-28 15:35:10.424502 +0300 +03  ping db succeeded, getdura:  21µs  execdur:  94.886ms
2024-04-28 15:35:15.521778 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  96.694ms
2024-04-28 15:35:20.6209 +0300 +03  ping db succeeded, getdura:  8µs  execdur:  98.624ms
2024-04-28 15:35:25.721061 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  98.833ms
2024-04-28 15:35:30.856921 +0300 +03  ping db succeeded, getdura:  55µs  execdur:  134.451ms
2024-04-28 15:35:35.954974 +0300 +03  ping db succeeded, getdura:  23µs  execdur:  97.512ms
2024-04-28 15:35:41.049324 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  92.95ms
2024-04-28 15:35:46.14955 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  98.742ms
2024-04-28 15:35:51.246783 +0300 +03  ping db succeeded, getdura:  58µs  execdur:  96.099ms
2024-04-28 15:35:56.33884 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  91.394ms
2024-04-28 15:36:01.441155 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  102.118ms
```

**Comment by sijms** [comment]
after setting `timeout=5`
```golang
2024-04-28 20:41:06.973845 +0300 +03  ping db succeeded, getdura:  663.595ms  execdur:  67.603ms
2024-04-28 20:41:12.048339 +0300 +03  ping db succeeded, getdura:  26µs  execdur:  73.874ms
2024-04-28 20:41:17.140386 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  90.811ms
2024-04-28 20:41:22.205758 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  64.161ms
2024-04-28 20:41:27.278699 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  71.883ms
2024-04-28 20:41:32.352729 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  72.821ms
2024-04-28 20:41:37.420044 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  66.082ms
2024-04-28 20:41:42.489153 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  67.869ms
2024-04-28 20:41:47.556788 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  67.139ms
2024-04-28 20:41:52.625404 +0300 +03  ping db succeeded, getdura:  5µs  execdur:  68.234ms
2024-04-28 20:41:57.715214 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  88.633ms
2024-04-28 20:42:02.801801 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  85.607ms
2024-04-28 20:42:07.876024 +0300 +03  ping db succeeded, getdura:  4µs  execdur:  73.037ms
2024-04-28 20:42:12.945687 +0300 +03  ping db succeeded, getdura:  6µs  execdur:  68.559ms
```

**Comment by freeNestor** [comment]
> you use `Connection timeout = 60` and `Timeout = 5` they are same parameter

How to achieve different timeout about connect timeout and session read timeout?
BTW, the key is that it returns error before context timeout period from my test output, and 2.8.11 is ok in the same environment, this confuses me.

**Comment by sijms** [comment]
the code after format:
```golang
// context.Background() as the parent. 
ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second) 
defer cancel() 
// Use this when testing the connection pool. 
if err = dbh.PingContext(ctx); err != nil { 
	log.Debugf("DB Connect returned error= %s", err) return nil, err 
}

// This works fine. Afterwards i am creating a prepared Query

// PrepareSQL parses a sql for a given connection and returns statement handler 
func PrepareSQL(dbh *sqlx.DB, mySQL string) (stmt *sqlx.Stmt, err error) { 
	log.Debugf("PrepareSql entered") 
	ok, t := common.CheckType(dbh, "*sqlx.DB") 
	if !ok { 
		err = fmt.Errorf("invalid dbh %s", t) 
		return 
	} 
	stmt, err = dbh.Preparex(mySQL) 
	if err != nil { 
		err = fmt.Errorf("prepare failed:%s (%s)", err, mySQL) 
	} 
	return 
}
// and execute it here
func SelectStmt(stmt *sqlx.Stmt, args ...any) (rows *sqlx.Rows, err error) { 
	log.Debugf("QueryStmt entered") 
	ok, t := common.CheckType(stmt, "*sqlx.Stmt") 
	if !ok { 
		err = fmt.Errorf("invalid stmt %s", t) 
		return 
	} 
	log.Debugf("Parameter values: %v", args...) 
	rows, err = stmt.Queryx(args...) 
	return 
}
```

**Comment by sijms** [comment]
I find the cause of the issue

**Comment by sijms** [comment]
you can test last commit for confirmation

**Comment by freeNestor** [comment]
ok, I'll test it on my first working day and back here.

**Comment by freeNestor** [comment]
test latest commit several times, this error disappears.

**Comment by sijms** [comment]
fixed 2.8.16

**Comment by Tommi2Day** [comment]
i can confirm rel 2.8.16 solved my issue
